Add docs for the App Builder database library (#457)
AdobeDocs/app-builder: src/pages/guides/app_builder_guides/storage/database.md

---
keywords:
  - Adobe I/O
  - App Builder
  - Database Storage
  - aio-lib-db
  - DocumentDB
  - MongoDB
title: Database Storage Getting Started
---

> # Overview

Reviewer: Title should match H1

There is a strict one-to-one relationship between an AIO Project Workspace and a Workspace Database, and each Workspace Database is entirely isolated from all other Workspace Databases.

Before use, a Workspace Database must be explicitly provisioned using either **aio-lib-db** or the AIO CLI. This is a self-service operation subject to organizational quotas and limits.

Workspace Databases are provisioned in one and only one of the following regions: `amer`, `apac`, and `emea`. These acronyms are defined as follows:

- `amer`: North, Central, and South America. Data is stored in the US.
- `apac`: Asia and Pacific. Data is stored in Japan.
- `emea`: Europe, the Middle East, and Africa. Data is stored in the EU.

## DB plugin in the AIO CLI

The DB Plugin in the AIO CLI is a utility that facilitates provisioning, initializing, querying, and monitoring Workspace Databases.

The following is only a brief introduction to the DB Plugin. For more thorough documentation see [aio-cli-plugin-app-storage](https://github.com/adobe/aio-cli-plugin-app-storage/tree/epic/abdb-implementation).

### Installation

> To be included in aio --version >= 11.0.0

Reviewer: Could link to or reproduce the aio install command

Reviewer 2: Not sure this should be included here, but prior to GA the following command is needed to install the db plugin:
```
aio plugins:install @adobe/aio-cli-plugin-app-storage@next
```

### Region selection

When using the DB Plugin in the AIO CLI, it is important that the region is the same as where the database is provisioned. If not, the connection will fail.

The default region is `amer`, and to set a different one, you can use the `--region` flag or add the `AIO_DB_REGION` variable to your .env file. Supported regions include `amer`, `emea`, and `apac`.

### Provisioning a workspace database

To provision a Workspace Database in the current AIO Project Workspace is as simple as:

```bash
aio app db provision
```

The provisioning status can be retrieved with this:

```bash
aio app db status
```

To check connectivity with the database:

```bash
aio app db ping
```

Additional database commands include:

```bash
aio app db stats               # Get statistics about your App Builder database
```

```bash

> aio app db delete              # Delete the database for your App Builder application (non-production only)

Reviewer: maybe make the descriptions their own line or take descriptions out of commands, so that they are more easily copied

```

### Collections

> Collections do not have to be explicitly created in order to start using them. But if specific fields need to be indexed or documents require schema validation, then creating a collection beforehand would make sense.

Reviewer: ```suggestion
Collections do not have to be explicitly created in order to start using them, but if specific fields need to be indexed or documents require schema validation, then creating a collection beforehand makes sense.
```

const user = await userCollection.findOne({ email: 'john@example.com' });
```

Find all documents matching a filter (returns a cursor - see next section):

```javascript
const cursor = userCollection.find({ age: { $gte: 18 } })
```

Find with projection, sort, skip and limit (returns a cursor - see next section):

```javascript
const cursor = userCollection.find({ age: { $gte: 18 } })
  .project({ name: 1, email: 1 })
  .sort({ name: 1 })
  .skip(2)
  .limit(10)
```

MongoDB Node Driver references:

- [Find Documents](https://www.mongodb.com/docs/drivers/node/current/crud/query/retrieve/)
- [Specify Documents to Return](https://www.mongodb.com/docs/drivers/node/current/crud/query/specify-documents-to-return/)
- [Specify Which Fields to Return](https://www.mongodb.com/docs/drivers/node/current/crud/query/project/)
- [Specify a Query](https://www.mongodb.com/docs/drivers/node/current/crud/query/query-document/)
- [Count Documents](https://www.mongodb.com/docs/drivers/node/current/crud/query/count/)
- [Retrieve Distinct Values](https://www.mongodb.com/docs/drivers/node/current/crud/query/distinct/)
- [Search Text](https://www.mongodb.com/docs/drivers/node/current/crud/query/text/)

> - [Search Geospatially](https://www.mongodb.com/docs/drivers/node/current/crud/query/geo/)

Reviewer: ```suggestion
- [Find documents](https://www.mongodb.com/docs/drivers/node/current/crud/query/retrieve/)
- [Specify documents to return](https://www.mongodb.com/docs/drivers/node/current/crud/query/specify-documents-to-return/)
- [Specify which fields to return](https://www.mongodb.com/docs/drivers/node/current/crud/query/project/)
- [Specify a auery](https://www.mongodb.com/docs/drivers/node/current/crud/query/query-document/)
- [Count documents](https://www.mongodb.com/docs/drivers/node/current/crud/query/count/)
- [Retrieve distinct values](https://www.mongodb.com/docs/drivers/node/current/crud/query/distinct/)
- [Search text](https://www.mongodb.com/docs/drivers/node/current/crud/query/text/)
- [Search geospatially](https://www.mongodb.com/docs/drivers/node/current/crud/query/geo/)
```

#### Cursor access patterns

Both **find** and **aggregate** return cursors.

Using toArray() - loads all results into memory:

```javascript
const results = await cursor.toArray()
```

Using iteration - memory efficient:

```javascript
while (await cursor.hasNext()) {
  const doc = await cursor.next();
  console.log(doc)
}
```

Using for await...of - most convenient:

```javascript
for await (const doc of cursor) {
  console.log(doc)
}
```

Using streams:

```javascript
const stream = cursor.stream();
stream.on('data', (doc) => {
  console.log(doc)
})
```

MongoDB Node Driver references:

> - [Access Data From a Cursor](https://www.mongodb.com/docs/drivers/node/current/crud/query/cursor/)

Reviewer: ```suggestion
- [Access data from a cursor](https://www.mongodb.com/docs/drivers/node/current/crud/query/cursor/)
```

#### Updating documents

Update one document:

```javascript
const result = await userCollection.updateOne(
  { email: 'john@example.com' },
  { $set: { age: 31 } }
)
```

Replace one document:

```javascript
const result = await userCollection.replaceOne(
  { email: 'john@example.com' },
  { name: 'Bob', email: 'bob@example.com', age: 12 }
)
```

Update multiple documents:

```javascript
const result = await userCollection.updateMany(
  { age: { $lt: 18 } },
  { $set: { category: 'minor' } }
)
```

Find and update:

```javascript
const updatedUser = await userCollection.findOneAndUpdate(
  { email: 'john@example.com' },
  { $set: { lastLogin: new Date() } },
  { returnDocument: 'after' }
)
```

MongoDB Node Driver references:

- [Modify Documents](https://www.mongodb.com/docs/drivers/node/current/crud/update/modify/)
- [Replace Documents](https://www.mongodb.com/docs/drivers/node/current/crud/update/replace/)

> - [Update Arrays in a Document](https://www.mongodb.com/docs/drivers/node/current/crud/update/embedded-arrays/)

Reviewer: Caps? Maybe another repo specific thing

- hasNext()
- next()
- toArray()
- close()

AWS reference: [Cursor Methods](https://docs.aws.amazon.com/documentdb/latest/developerguide/mongo-apis.html#mongo-apis-cursor)

### Aggregation pipeline operators

At this time only the following stage operators are supported by App Builder Database:

- $addFields
- $count
- $geoNear
- $group
- $limit
- $lookup
- $match
- $out
- $project
- $redact
- $replaceRoot
- $sample
- $skip
- $sort
- $unwind

> **Note:** Support for more stage operators is coming soon.

All other pipeline operators are the same as for DocumentDB 8.0.

AWS reference: [Aggregation pipeline operators](https://docs.aws.amazon.com/documentdb/latest/developerguide/mongo-apis.html#mongo-apis-aggregation-pipeline)

### Data types

Same as DocumentDB 8.0.

AWS reference: [Data types](https://docs.aws.amazon.com/documentdb/latest/developerguide/mongo-apis.html#mongo-apis-data-types)

### Indexes and index properties

Same as DocumentDB 8.0.

AWS reference: [Indexes and index properties](https://docs.aws.amazon.com/documentdb/latest/developerguide/mongo-apis.html#mongo-apis-index)

## Support and contact information

> For any questions, issues, or feedback regarding App Builder Database Storage, please reach out to the team at:

Reviewer: ```suggestion
For any questions, issues, or feedback regarding App Builder Database Storage, reach out to the team using one of the following:
```
